tools/test_read_only.sh

```diff
@@ -0,0 +1,202 @@
+# Loop over 2 missing downstairs, activate and verify our volume.
+set -o pipefail
+SECONDS=0
+
+ROOT=$(cd "$(dirname "$0")/.." && pwd)
+BINDIR=${BINDIR:-$ROOT/target/debug}
+
+echo "$ROOT"
+cd "$ROOT" || (echo failed to cd "$ROOT"; exit 1)
+
+if pgrep -fl crucible-downstairs; then
+    echo 'Downstairs already running?' >&2
+    exit 1
+fi
+
+cds="$BINDIR/crucible-downstairs"
+crutest="$BINDIR/crutest"
+dsc="$BINDIR/dsc"
+for bin in $cds $crutest $dsc; do
+    if [[ ! -f "$bin" ]]; then
+        echo "Can't find crucible binary at $bin" >&2
+        exit 1
+    fi
+done
+
+# Use the id of the current user to make a unique path
+user=$(id -u -n)
+# Downstairs regions go in this directory
+testdir="/var/tmp/test_read_only-$user"
+if [[ -d "$testdir" ]]; then
+    rm -rf "$testdir"
+fi
+
+# Store log files we want to keep in /tmp/test_read_only/.txt as this is what
+# buildomat will look for and archive
+test_output_dir="/tmp/test_read_only-$user"
+rm -rf "$test_output_dir" 2> /dev/null
+mkdir -p "$test_output_dir"
+log_prefix="${test_output_dir}/test_read_only"
+fail_log="${log_prefix}_fail.txt"
+rm -f "$fail_log"
+verify_file="${log_prefix}_verify"
+rm -f "$verify_file"
+test_log="${log_prefix}_log"
+rm -f "$test_log"
+dsc_output_dir="${test_output_dir}/dsc"
+mkdir -p "$dsc_output_dir"
+dsc_output="${test_output_dir}/dsc-out.txt"
+echo "dsc output goes to $dsc_output"
+
+region_count=3
+args=()
+dsc_args=()
+dsc_create_args=()
+upstairs_key=$(openssl rand -base64 32)
+echo "Upstairs using key: $upstairs_key" | tee -a "$test_log"
+
+args+=( --key "$upstairs_key" )
+args+=( --dsc "127.0.0.1:9998" )
+dsc_create_args+=( --encrypted )
+dsc_create_args+=( --cleanup )
+dsc_args+=( --output-dir "$dsc_output_dir" )
+dsc_args+=( --ds-bin "$cds" )
+dsc_args+=( --region-dir "$testdir" )
+
+# Control-C to cleanup.
+trap ctrl_c INT
+function ctrl_c() {
+	echo "Stopping at your request" | tee -a "$test_log"
+    ${dsc} cmd shutdown
+    exit 1
+}
+
+echo "Creating $region_count downstairs regions"
+echo "${dsc}" create "${dsc_create_args[@]}" --region-count "$region_count" --extent-size 10 --extent-count 5 "${dsc_args[@]}" > "$dsc_output"
+"${dsc}" create "${dsc_create_args[@]}" --region-count "$region_count" --extent-size 10 --extent-count 5 "${dsc_args[@]}" >> "$dsc_output" 2>&1
+
+echo "Starting $region_count downstairs"
+echo "${dsc}" start "${dsc_args[@]}" --region-count "$region_count" >> "$dsc_output"
+"${dsc}" start "${dsc_args[@]}" --region-count "$region_count" >> "$dsc_output" 2>&1 &
+dsc_pid=$!
+echo "dsc started at PID: $dsc_pid"
+
+while "$dsc" cmd all-running | grep false > /dev/null ; do
+    echo "Wait for all clients to be running."
+    sleep 3
+done
+
+echo ""
+echo "Begin tests, output goes to $test_log"
+gen=1
+
+# Initial fill and creation of verify file for future use.
+echo "$crutest" generic -g $gen "${args[@]}" --stable --verify-out "$verify_file" -c 250 | tee -a "$test_log"
```

Comment:
Why does this line specifically use `tee -a`, versus `>>` for other "log the command before running it" lines?

Reply:
It was my personal preference based on yesterdays mood on what I wanted to look at while watching the test run. :)

I, somewhat arbitrarily decided I only wanted to see the summary of what the test is doing and not the full crutest output, but I do want to log everything so there is a full record if there is a problem.